src/signHash.c

```diff
@@ -0,0 +1,290 @@
+	switch (button_mask) {
+	// The available button mask values are LEFT, RIGHT, EVT_RELEASED, and
+	// EVT_FAST. EVT_FAST is set when a button is held for 8 "ticks," i.e.
+	// 800ms.
+	//
+	// The comparison screens in the Zilliqa app allow the user to scroll using
+	// the left and right buttons. The user should be able to hold a button
+	// and scroll at a constant rate. When the user first presses the left
+	// button, we'll hit the LEFT case; after they've held the button for 8
+	// ticks, we'll hit the EVT_FAST | LEFT case. Since we want to scroll at a
+	// constant rate regardless, we handle both cases identically.
+	//
+	// Also note that, unlike the approval screen, we don't check for
+	// EVT_RELEASED. In fact, when a single button is released, none of the
+	// switch cases will be hit, so we'll stop scrolling.
+	case BUTTON_LEFT:
+	case BUTTON_EVT_FAST | BUTTON_LEFT: // SEEK LEFT
+		// Decrement the displayIndex when the left button is pressed (or held).
+		if (ctx->displayIndex > 0) {
+			ctx->displayIndex--;
+		}
+		// Use the displayIndex to recalculate the displayed portion of the
+		// text. os_memmove is the Ledger SDK's version of memmove (there is
+		// no os_memcpy). In practice, I don't think it matters whether you
+		// use os_memmove or the standard memmove from <string.h>.
+		os_memmove(ctx->partialHashStr, ctx->hexHash+ctx->displayIndex, 12);
+		// Re-render the screen.
+		UX_REDISPLAY();
+		break;
+
+	case BUTTON_RIGHT:
+	case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
+		if (ctx->displayIndex < sizeof(ctx->hexHash)-12) {
+			ctx->displayIndex++;
+		}
+		os_memmove(ctx->partialHashStr, ctx->hexHash+ctx->displayIndex, 12);
+		UX_REDISPLAY();
+		break;
+
+	case BUTTON_EVT_RELEASED | BUTTON_LEFT | BUTTON_RIGHT: // PROCEED
+		// Prepare to display the approval screen by printing the key index
+		// into the indexStr buffer. We copy two bytes in the final os_memmove
+		// so as to include the terminating '\0' byte for the string.
+		os_memmove(ctx->indexStr, "with Key #", 10);
+		int n = bin64b2dec(ctx->indexStr+10, sizeof(ctx->indexStr)-10, ctx->keyIndex);
+		os_memmove(ctx->indexStr+10+n, "?", 2);
+		// Note that because the approval screen does not have a preprocessor,
+		// we must pass NULL.
+		UX_DISPLAY(ui_signHash_approve, NULL);
+		break;
+	}
+	// (The return value of a button handler is irrelevant; it is never
+	// checked.)
+	return 0;
+}
+
+// handleSignHash is the entry point for the signHash command. Like all
+// command handlers, it is responsible for reading command data from
+// dataBuffer, initializing the command context, and displaying the first
+// screen of the command.
+void handleSignHash(uint8_t p1, uint8_t p2, uint8_t *dataBuffer, uint16_t dataLength, volatile unsigned int *flags, volatile unsigned int *tx) {
+	// Read the index of the signing key. U4LE is a helper macro for
+	// converting a 4-byte buffer to a uint32_t.
+	ctx->keyIndex = U4LE(dataBuffer, 0);
```

Comment:
You know the length of the input buffer (dataLength).
I'd add the following check to avoid reading uninitialized data in databuffer:

```c
if (dataLength != sizeof(uint32_t) + sizeof(ctx->hash)) {
  // THROW or return something
}